Log improvements

davidrissato opened the pull request:
Added two new log methods to log.py and some extra log event to pipleline_flow and pipeline_step.

With this change, logging and measuring execution of code fragments is much easier.
Example:
```python
# some_file.py
from log import log_context

def some_method
    # ... some code here ...
    with log_context("sub_step", {"abc": "123"}):
        # ... your code fragment goes here ...
```
The code above will generate two log entries before and after executing the code fragment:
```
Event ctx_start {"n": "sub_step", "v": {abc": 123}, "caller": {"f": ["some_file.py", "some_method"]}
Event ctx_end {"n": "sub_step", "v": {abc": 123}, "caller": {"f": ["some_file.py", "some_method"]} (0.3 sec)
```

tfrcarvalho commented on idseq_dag/util/log.py:
```diff
@@ -34,6 +39,70 @@ def write(message, warning=False, flush=True):
+        val["error_type"] = type(e).__name__
+        val["error_args"] = e.args
+        log_event("ctx_error", val, start_time=start)
+        raise e
```
This is one of those occasions where I do not agree with re-raising. 
Logging should be itself a parallel and non-intrusive action to the main logic.

davidrissato replied:
This re-raise is exactly there to be non-intrusive. This catch block is catching an error on the wrapped method and re-raising the same error.

tfrcarvalho replied:
I am just saying that it is not responsability of a log util to do that.
Here you should catch and handle errors on the log function and not the others.
Also, I believe this will add the log to the stack trace, adding to the confusion. Can you confirm that?

davidrissato replied:
@tfrcarvalho :
It keeps the same info:
![image](https://user-images.githubusercontent.com/8084076/56986302-3fdbec80-6b3f-11e9-8153-1ab5ca5770c6.png)
```python
import traceback

from contextlib import contextmanager

@contextmanager
def log_context():
    print("ctx_start")
    try:
        yield
        print("ctx_end")
    except Exception as e:
        print("ctx_error")
        raise e

def throws_error():
  1 / 0 # Division by zero

def test1():
  try:
    throws_error()
  except:
    traceback.print_exc()

def test2():
  try:
    with log_context():
      throws_error()
  except:
    traceback.print_exc()

print("-------- test1() stacktrace --------")
test1()
print();
print("-------- test2() stacktrace --------")
test2()
```

davidrissato replied:
The idea here is not hiding errors or even use this feature to replace the original error handlers. You will notice that I'm not even printing the stack trace, because I feel this is a task for the caller.
The idea of this log context is to provide a way to know when a code block started and ended, and how long it took. The problem is if we don't catch exceptions here (and re-raise them), we will never know if the code has failed or if it has never completed.

davidrissato replied:
The new  PR `log_context` provides a way to know when a code block started and ended, and how long it takes. The problem is if we don't catch exceptions here (and re-raise them), we will never know if the code has failed or if it has never completed.
The idea is not hiding errors or even use this feature to replace the original error handlers. You will notice that I'm not even printing the stack trace, because I feel this is a task for the caller.

tfrcarvalho replied:
I think it is fine then.

tfrcarvalho commented on idseq_dag/util/log.py:
```diff
@@ -34,6 +39,70 @@ def write(message, warning=False, flush=True):
+    start = log_event("ctx_start", val)
+    try:
+        yield
+        log_event("ctx_end", val, start_time=start)
```
I would add elapsed time here. Makes it easier to analyze on the logs.

tfrcarvalho replied:
Nevermind, saw you are doing that above.

tfrcarvalho commented on idseq_dag/util/log.py:
```diff
@@ -34,6 +39,70 @@ def write(message, warning=False, flush=True):
+        fmt_message = "Event '%s' %s" % (event_name, fmt_values)
+    else:
+        duration = (datetime.datetime.now()-start_time).total_seconds()
+        fmt_message = "Event '%s' %s (%.1f seconds)" % (event_name, fmt_values, duration)
```
We were considering making all log messages in json format for easier parsing downstream.

davidrissato replied:
About JSON: this is super easy to change, but I haven't completely done that yet because there are a lot of "non-json" logs floating around. How does Datadog handle log files that mix JSON and text?

tfrcarvalho replied:
Not sure. You might want to sync with @jshoe when he comes back.
I think Datadog will still index the free text, but will recognize everything as the same field. 
You can either start changing to JSON or wait to sync with Jonathan and understand Datadog better to do that. I would go with the first option but is up to you.

davidrissato replied:
I feel like going to full JSON might be part of a separate PR. We probably need to define some additional standards about what fields will be part of the json in which format, name, (ex: timestamp).

yunfangjuan commented on idseq_dag/util/log.py:
```diff
@@ -34,6 +39,70 @@ def write(message, warning=False, flush=True):
+@contextmanager
+def log_context(context_name, values=None, log_caller_info=True):
+    '''
+    Log context manager to track started and completed events for a block of code
```
nice

davidrissato commented:
The idea here is not hiding errors or even use this feature to replace the original error handlers. You will notice that I'm not even printing the stack trace, because I feel this is a task for the caller.
The idea of this log context is to provide a way to know when a code block started and ended, and how long it took. The problem is if we don't catch exceptions here (and re-raise them), we will never know if the code has failed or if it has never completed.

davidrissato commented:
About JSON: this is super easy to change, but I haven't completely done that yet because there are a lot of "non-json" logs floating around. How does Datadog handle log files that mix JSON and text?

tfrcarvalho commented on idseq_dag/util/log.py:
```diff
@@ -34,6 +39,70 @@ def write(message, warning=False, flush=True):
+            # ... your code goes here ...
+
+    The code above will generate two log entries:
+    Event ctx_start {"n": "sub_step", "v": {abc": 123}, "caller": {"f": ["some_file.py", "some_method"]}
```
Why `n`, `v` vs. readable names? Just to save space?

davidrissato replied:
No, it was more related to declutter the visualization.
I had originally called them `context_name` and `values`, but I felt it created a lot of noise when reading these lines through CloudFront, which is our main tool to read these logs currently. Long names certainly wouldn't be a problem if we were fully ingesting these lines in JSON format and displaying in tabular format.

I also enjoy using descriptive names and I'm don't have a strong position on keeping them using short names. I'm very open to change it.

But I wanted a second opinion on how cluttered it actually gets.
Let me show an example:

### using short (and unreadable) names
```
2019-04-30 20:18:25,205: [    Thread-4] Command 9: date > /mnt/idseq/results/10589/banana_input.txt.done
2019-04-30 20:18:25,210: [    Thread-4] Event 'ctx_end' {"n": "fetch_input_files_from_s3", "v": {"target": "banana_in"}, "caller": {"f": ["pipeline_flow.py", "fetch_target_from_s3"]}} (1.4 seconds)
2019-04-30 20:18:28,849: [   Thread-11] Event 'ctx_end' {"n": "substep_wait_for_input_files", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}} (5.0 seconds)
2019-04-30 20:18:28,850: [   Thread-11] Event 'ctx_start' {"n": "substep_run", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,850: [   Thread-11] BANANA
2019-04-30 20:18:28,851: [   Thread-11] Event 'ctx_end' {"n": "substep_run", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,851: [   Thread-11] Event 'ctx_start' {"n": "substep_validate", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,852: [   Thread-11] Command 10: date > /mnt/idseq/results/10589/banana1.txt.done
2019-04-30 20:18:28,858: [   Thread-11] Command 11: date > /mnt/idseq/results/10589/banana2.txt.done
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_end' {"n": "substep_validate", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_start' {"n": "substep_save_progress", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_end' {"n": "substep_save_progress", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_start' {"n": "substep_save_counts", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,863: [   Thread-11] Event 'ctx_end' {"n": "substep_save_counts", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,863: [   Thread-11] Event 'ctx_end' {"n": "dag_step", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}} (5.0 seconds)
2019-04-30 20:18:28,863: [   Thread-16] Command 12: aws s3 cp --only-show-errors /mnt/idseq/results/10589/banana1.txt s3://sample-bucket/david/test_001/out/3.5/banana1.txt
2019-04-30 20:18:29,499: [   Thread-16] Command 13: aws s3 cp --only-show-errors /mnt/idseq/results/10589/banana2.txt s3://sample-bucket/david/test_001/out/3.5/banana2.txt
2019-04-30 20:18:30,173: [  MainThread] all steps are done
```

**vs**

### using descriptive names
```
2019-04-30 20:18:25,205: [    Thread-4] Command 9: date > /mnt/idseq/results/10589/banana_input.txt.done
2019-04-30 20:18:25,210: [    Thread-4] Event 'ctx_end' {"context_name":"fetch_input_files_from_s3", "values":{"target": "banana_in"}, "caller": {"function":["pipeline_flow.py", "fetch_target_from_s3"]}} (1.4 seconds)
2019-04-30 20:18:28,849: [   Thread-11] Event 'ctx_end' {"context_name":"substep_wait_for_input_files", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}} (5.0 seconds)
2019-04-30 20:18:28,850: [   Thread-11] Event 'ctx_start' {"context_name":"substep_run", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,850: [   Thread-11] BANANA
2019-04-30 20:18:28,851: [   Thread-11] Event 'ctx_end' {"context_name":"substep_run", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,851: [   Thread-11] Event 'ctx_start' {"context_name":"substep_validate", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,852: [   Thread-11] Command 10: date > /mnt/idseq/results/10589/banana1.txt.done
2019-04-30 20:18:28,858: [   Thread-11] Command 11: date > /mnt/idseq/results/10589/banana2.txt.done
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_end' {"context_name":"substep_validate", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_start' {"context_name":"substep_save_progress", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_end' {"context_name":"substep_save_progress", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_start' {"context_name":"substep_save_counts", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,863: [   Thread-11] Event 'ctx_end' {"context_name":"substep_save_counts", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,863: [   Thread-11] Event 'ctx_end' {"context_name":"dag_step", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}} (5.0 seconds)
2019-04-30 20:18:28,863: [   Thread-16] Command 12: aws s3 cp --only-show-errors /mnt/idseq/results/10589/banana1.txt s3://sample-bucket/david/test_001/out/3.5/banana1.txt
2019-04-30 20:18:29,499: [   Thread-16] Command 13: aws s3 cp --only-show-errors /mnt/idseq/results/10589/banana2.txt s3://sample-bucket/david/test_001/out/3.5/banana2.txt
2019-04-30 20:18:30,173: [  MainThread] all steps are done
```

tfrcarvalho replied:
I like the second one, and think it will be better going forward and as we start analyzing these in data dog.

tfrcarvalho commented on idseq_dag/util/log.py:
```diff
@@ -34,6 +39,70 @@ def write(message, warning=False, flush=True):
+        val["v"] = values
+    if log_caller_info:
+        f_code = sys._getframe(2).f_code
+        val["caller"] = {"f": [os.path.basename(f_code.co_filename), f_code.co_name]}
```
Why f and not directly `val["caller"] = [os.path.basename(f_code.co_filename), f_code.co_name]`. Are you planning on saving something else?

davidrissato replied:
Yes, I wanted to leave room for future extensions, for example, line number.

tfrcarvalho replied:
IMO, in that case, i would just use an hash like `{"caller": {filename: ..., name: ..., lineno: }` instead of the array...

davidrissato commented:
Added a couple changes:
- Using readable attributes instead of `v`, `f` etc.
- Converted the entire event to a JSON object. Easier to parse on Datadog.
- Added field `duration_ms` (duration in milliseconds) to this new json object

davidrissato commented:
I just realized that I should probably have changed `__init__.py` and `read.me` to bump version. Doing that now

davidrissato commented:
Well, that should be a non-invasive change (except for one log entry that got replace by the new mechanism), and I also tested that in staging by running a pipeline pointing to my modified version, but I believe next time it would be better adding a unit test to it.

cdebourcy commented:
Great. I was actually just looking for this "I also tested that in staging by running a pipeline pointing to my modified version" -- helpful for the reviewer for any PR in general :)

cdebourcy commented:
I saw that your sample in staging was marked FAILED (and I believe that was the intended outcome) -- did you try getting a successful run as well?

davidrissato commented:
Oh, now I understood what you are asking.

Considering this is just a log change, every other log line that hasn't failed is a test for success :-)

Actually, I created my own step and exhaustively tested it locally exercising both failing and success scenarios. Let me add a new run (sorry, I haven't captured previous ones):

**Success**:
![image](https://user-images.githubusercontent.com/8084076/57036279-b09b0b80-6c08-11e9-95d7-3234b2a3a095.png)

**Error**:
![image](https://user-images.githubusercontent.com/8084076/57036319-d0caca80-6c08-11e9-9aa6-b1334fe0c0be.png)

davidrissato commented:
I'm adding that to the PR description. Now I realized that I missed the PR template for this project like we have on `idseq_web`.
